src/lib/ts/cache-compiler-host.ts

```diff
@@ -289,3 +288,15 @@ export function augmentProgramWithVersioning(program: ts.Program): void {
     return files;
   };
 }
+
+export function invalidateProgramFiles(program: ts.Program, files: Set<string>): void {
```

Comment:
It's odd, since TypeScript should manage this automatically. I'm fairly sure we don’t handle this in the Angular CLI.

Could you provide a detailed explanation or reproduction steps to show why this is needed?

//cc @JoostK, as he has a deep understanding of TypeScript internals.

Reply:
I'd have to check some things to determine what is the case here. Is there a test that would show the issue and how these changes address it?

Reply:
I noticed that when you save a component file for a couple of times that as errors in a template, suddenly, the diagnostics stop being reported and the entry points gets built normally.

I think the problem is that we are relying on the `affectedFiles` to run the angular diagnostics, and when you re-save a file without changing its content, that file is not marked as affected anymore by the `getSemanticDiagnosticsOfNextAffectedFile`. 

`compile-source-files.ts`
```ts
// Only request Angular template diagnostics for affected files to avoid
// overhead of template diagnostics for unchanged files.
if (affectedFiles.has(sourceFile)) {
  const angularDiagnostics = angularCompiler.getDiagnosticsForFile(...);
  ...
}
```

I was making sure those files are re-versioned so they are picked up again by the builder but maybe I could change that condition in some way?

On a final note, in the angular compiler cli, the builder is not reused again when there are errors, I think.

Reply:
> I noticed that when you save a component file for a couple of times that as errors in a template, suddenly, the diagnostics stop being reported and the entry points gets built normally.

Oh that is expected and working as intended from TS point of view.  we should be storing and reusing previous errors and failures instead of invalidating the file unless it has actually changed.

> On a final note, in the angular compiler cli, the builder is not reused again when there are errors, I think.

The compiler CLI does not utilize TypeScript builders, and it just re-uses the old program.

Reply:
Ok that makes sense, I will update my PR then. I am thinking in comparing the version of the previous affected files with the version of the matching current files, and if they are the same, I will just retrieve the old diagnostics.

Reply:
I have updated the code. I am not wasting your time. :(